### ISSUE TITLE
Custom entry field support for citation intellisense

### ISSUE THREAD

**Issue by nerdunio** [James-Yu/LaTeX-Workshop#3214]
I found `latex-workshop.intellisense.citation.label` and `latex-workshop.intellisense.citation.type` to customize the citation completion. I set the type to "browser" to see more information about the entry. This is a nice feature.

However, it would be great if I could define a list of entry fields instead of a predefined string, and there's why.

In my bib files I use custom entry fields, for example "description". This field is like the title for some of my entries. With the current selection of "authors", "title" or "bibtex key" I can't target custom fields nor two fields simultaneously. 

So I suggest to extend the label from a string to an array, just like `latex-workshop.intellisense.citation.format`:

`"latex-workshop.intellisense.citation.label" : ["title", "description"],`

Hence, the two entries

```
@book{lamport1994latex,
  title={LATEX: A Document Preparation System : User's Guide and Reference Manual},
  author={Lamport, L. and Bibby, D. and Pearson Education},
  isbn={9780201529838},
  lccn={93039691},
  series={Addison-Wesley Series on Tools},
  url={https://books.google.ch/books?id=khVUAAAAMAAJ},
  year={1994},
  publisher={Addison-Wesley}
}

@book{lamport1994latex:p31,
  description={Helpful hints},
  page={31},
}
```

would result in the following suggestions.

```
LATEX: A Document Preparation System : User's Guide and Reference Manual
Helpful hints
```

And then I can type "latex guide" or "hints" and get one of the above suggestions.

Until how I used LaTeXTools for Sublime text. There I was able to customize the completion as follows:

`"cite_panel_format": ["{title} - {description}, ({keyword})","{author_short}"],`

I've gotten so used to this feature that I can't work efficiently without it.

I really like LaTeX Workshop so far and this feature would make it even better :)

Thanks in advance!

**Comment by jlelong** [comment]
I am afraid this is not so easy. The setting `latex-workshop.intellisense.citation.label` must really be a single item because otherwise the rendering may become ugly: the label is displayed on a single line. 

Using the `inline` mode, you can display all the fields you want

<img width="792" alt="Capture d’écran 2022-03-22 à 17 46 36" src="https://user-images.githubusercontent.com/2910140/159531854-836cf38d-7c36-4b03-8fcb-a8cac21ef7f0.png">

I have just added `description` to `latex-workshop.intellisense.citation.format`. I hope that helps.

**Comment by nerdunio** [comment]
Thank you for your fast reply!

Yes, adding multiple fields to the label can lead to an ugly result. However, adding just the `title `and `description` wouldn't result in a mess.

About `inline` and `browser` mode: On my screen resolution it seems that the `browser` mode provides more space, even for more than one entry field on the first line, compared to the `inline`.
![browser_vs_inline_vscode_lw](https://user-images.githubusercontent.com/56926255/159538054-13bb5d44-f0e2-4b07-8269-5a9335e6b501.png)

And if I replace my `description` field with `title` I get the desired result. Thus, no ugly rendering:
![image](https://user-images.githubusercontent.com/56926255/159538333-a659619f-c00f-4c46-b522-970cf8b518da.png)

**Comment by nerdunio** [comment]
And, I was unable to display my `description `field in `inline ` mode over `format`.

**Comment by nerdunio** [comment]
Yes, that would be great!

**Comment by James-Yu** [comment]
I've got another idea to think. What about we change the type of `citation.label` from `string` to `list`, and consider the list as a priority list so that the first available field is displayed? I'm still not sure if template strings like `["{title} - {description}, ({keyword})","{author_short}"]` should be implemented, but priority list should be easy.

Compatibility: we may have a fallback check on the type of this config. If string, use it. If list, iteratively check (performance issue?).

Also, it seems that this is only concerned with `inline`. Do we need to improve (somehow) browser display?

**Comment by jlelong** [comment]
@James-Yu I actually made a POC last night to consider `latex-workshop.intellisense.citation.format` as the secondary display in browser mode and to use these fields as filtering text. 
@James-Yu @nerdunio Can you try #3216? See [here](https://github.com/jlelong/LaTeX-Workshop/blob/master/CONTRIBUTING.md) for how to test the extension

**Comment by nerdunio** [comment]
> > And, I was unable to display my `description `field in `inline ` mode over `format`.
> 
> I set
> 
> ```
> "latex-workshop.intellisense.citation.format": [
>     "author",
>     "title",
>     "journal",
>     "description",
>    "year"
> ]
> ```
> 
> To display the details of suggestion item, click on the small arrow at the end of the line. One you have enabled details once, they will keep showing up. Note the intellisense popup and the detail window are both resizable.

Got it, I didn't know that. Again a great feature! However, in this case I can't fuzzy search for "hints" and get the correct key of hundreds of entries.

**Comment by nerdunio** [comment]
> @James-Yu I actually made a POC last night to consider `latex-workshop.intellisense.citation.format` as the secondary display in browser mode and to use these fields as filtering text. @James-Yu @nerdunio Can you try #3216? See [here](https://github.com/jlelong/LaTeX-Workshop/blob/master/CONTRIBUTING.md) for how to test the extension

I will try it and come back. Thank you!

**Comment by nerdunio** [comment]
I followed the [instructions](https://github.com/jlelong/LaTeX-Workshop/blob/master/CONTRIBUTING.md) but it seems LW is still on an older version; the secondary display in the browser only shows `Authors` and `publication`.

![image](https://user-images.githubusercontent.com/56926255/159682483-ae4ff7a8-5717-474a-94d6-73c929941aea.png)

I start debug with "Run extension" configuration. I have all other extensions disabled (except `ESLint`) and I can see LW in the sidebar during debugging, so I assume it is running the latest version. 

https://github.com/James-Yu/LaTeX-Workshop/pull/3216 looks promising. Using the format as secondary display in browser mode is a great idea. One thing to note is that the browser now shows the title field twice.

@James-Yu the idea of the priority is great. This would solve the problem when using `inline` mode.

**Comment by jlelong** [comment]
> Also, it seems that this is only concerned with `inline`. Do we need to improve (somehow) browser display?
> @James-Yu the idea of the priority is great. This would solve the problem when using inline mode.

It is not clear to me yet what is exactly the issue with the inline mode. I am not really convinced with the priority list as it implies that for every bibliography entry we have to recompute which label to use and in the end we come up with a list of mixed labels. I think this is better to keep a fixed label. What use cases are not solved by #3216 ?

**Comment by nerdunio** [comment]
Now I was able to test the new feature.

Here's what I did wrong: I used the url from your [instructions](https://github.com/jlelong/LaTeX-Workshop/blob/master/CONTRIBUTING.md), that's why I cloned the James-Yu's repo instead of your forke with the correct branch.

In `browser` mode it works perfectly fine:
![image](https://user-images.githubusercontent.com/56926255/159719466-949c3ba4-ef58-4c55-b95a-9dbd5b66bd96.png)

and in `inline` mode too:
![image](https://user-images.githubusercontent.com/56926255/159719587-cc9087e3-1bd0-42ed-90fc-43fbed8126d4.png)

> What use cases are not solved by #3216 ?

Everything is solved by #3216. In `inline` mode it is important to click the arrow at the end of the line to see the fields!

Thank you very much!

**Comment by jlelong** [comment]
Great! I will continue to work on #3216 to come up with a true PR and not only a POC.

**Comment by nerdunio** [comment]
Nice, can't wait to pull and test the true PR!

**Comment by jlelong** [comment]
Closed by #3216
